src/business_logic/state/cached_state.rs

```diff
@@ -173,17 +176,35 @@ impl<T: StateReader + Clone> StateReader for CachedState<T> {
     fn get_compiled_class(
         &mut self,
         compiled_class_hash: &ClassHash,
-    ) -> Result<&CasmContractClass, StateError> {
-        let mut casm_class = self.get_casm_classes()?.clone();
-        if casm_class.get(compiled_class_hash).is_none() {
-            let casm = self
-                .state_reader
-                .get_compiled_class(compiled_class_hash)?
-                .clone();
-            casm_class.insert(*compiled_class_hash, casm);
-            self.casm_contract_classes = Some(casm_class);
+    ) -> Result<CompiledClass, StateError> {
+        if let Some(casm_class) = &mut self.casm_contract_classes {
+            if let Some(class) = casm_class.get(compiled_class_hash) {
+                return Ok(CompiledClass::Casm(Box::new(class.clone())));
+            }
+        }
+        if let Some(contract_class) = &mut self.contract_classes {
+            if let Some(class) = contract_class.get(compiled_class_hash) {
+                return Ok(CompiledClass::Deprecated(Box::new(class.clone())));
+            }
+        }
+        let contract = self.state_reader.get_compiled_class(compiled_class_hash);
+        match contract {
+            Ok(CompiledClass::Casm(class)) => {
+                if let Some(casm_class) = &mut self.casm_contract_classes {
+                    casm_class.insert(*compiled_class_hash, *class.clone());
+                    self.casm_contract_classes = Some(casm_class.clone());
+                }
+                Ok(CompiledClass::Casm(class))
+            }
+            Ok(CompiledClass::Deprecated(class)) => {
+                if let Some(contract_class) = &mut self.contract_classes {
+                    contract_class.insert(*compiled_class_hash, *class.clone());
+                    self.contract_classes = Some(contract_class.clone());
+                }
+                Ok(CompiledClass::Deprecated(class))
+            }
+            _ => Err(StateError::NoneCompiledClass(*compiled_class_hash)),
```

Comment:
This will be much much simpler when we get rid of contract_classes and casm_contract_classes hashmaps and only have compiled_classes hashmap